server/src/main/java/org/elasticsearch/action/admin/indices/mapping/get/TransportGetFieldMappingsIndexAction.java

```diff
@@ -81,11 +83,36 @@ protected GetFieldMappingsResponse shardOperation(final GetFieldMappingsIndexReq
         assert shardId != null;
         IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
         Predicate<String> metadataFieldPredicate = (f) -> indexService.mapperService().isMetadataField(f);
-        Predicate<String> fieldPredicate = metadataFieldPredicate.or(indicesService.getFieldFilter().apply(shardId.getIndexName()));
+        Predicate<String> baseFieldPredicate = metadataFieldPredicate.or(indicesService.getFieldFilter().apply(shardId.getIndexName()));
 
-        MappingLookup mappingLookup = indexService.mapperService().mappingLookup();
-        Map<String, FieldMappingMetadata> fieldMapping = findFieldMappings(fieldPredicate, mappingLookup, request);
-        return new GetFieldMappingsResponse(singletonMap(shardId.getIndexName(), fieldMapping));
+        DocumentMapper documentParser = indexService.mapperService().documentMapper();
+        if (documentParser == null) {
+            // no mappings
+            return new GetFieldMappingsResponse(singletonMap(shardId.getIndexName(), Collections.emptyMap()));
+        }
+
+        Predicate<String> fieldPredicate = fieldPredicate(baseFieldPredicate, request.fields());
+        ToXContent.Params params = request.includeDefaults()
+            ? new ToXContent.MapParams(Collections.singletonMap("include_defaults", "true"))
+            : ToXContent.EMPTY_PARAMS;
+
+        Map<String, FieldMappingMetadata> mappings = new HashMap<>();
+
+        documentParser.mapping().forEachMapper(m -> {
+            FieldMappingMetadata metadata = buildFieldMappingMetadata(fieldPredicate, m.name(), m, params);
+            if (metadata != null) {
+                mappings.put(metadata.fullName(), metadata);
+            }
+        });
+
+        for (RuntimeField rf : documentParser.mapping().runtimeFields()) {
```

Comment:
Is it an option to instead expose a single Map<String, ToXContent> ? It is not particularly pretty but it has two advantages: one method instead of two, and does not expose the distinction between runtime fields and concrete fields, which I am not sure we want and I would not want it to leak out without us being aware that that's what we want :)

Reply:
I thought about trying this, but it feels very odd to expose things as ToXContent from the mapping.  Doing it this way also allows us to be clear about shadowing (we pull mappers first, and then runtime fields for overrides), and I do think that making the distinction between the two makes sense in this context, where we're looking at mapping definitions themselves.